storage/local/frankenstein.go

```diff
@@ -259,3 +266,80 @@ func (i *Ingestor) Collect(ch chan<- prometheus.Metric) {
 	ch <- i.storedChunks
 	ch <- i.chunkStoreFailures
 }
+
+type invertedIndex struct {
+	mtx sync.RWMutex
+	idx map[model.LabelName]map[model.LabelValue][]model.Fingerprint // entries are sorted in fp order?
+}
+
+func newInvertedIndex() *invertedIndex {
+	return &invertedIndex{
+		idx: map[model.LabelName]map[model.LabelValue][]model.Fingerprint{},
+	}
+}
+
+func (i *invertedIndex) add(metric model.Metric, fp model.Fingerprint) {
```

Comment:
We'll also want a `del()` method which not only removes fingerprints, but also removes maps completely once the last contained fingerprint was removed. Otherwise we have a memory leak...